internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,360 @@
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Child("valueFrom", "variable"),
+						jsonPatch.ValueFrom.Variable,
+						fmt.Sprintf("variable with name %s cannot be found", *jsonPatch.ValueFrom.Variable),
+					))
+			}
+		}
+	}
+	return allErrs
+}
+
+// This contains a list of all of the valid builtin variables.
+// TODO(killianmuldoon): Match this list to controllers/topology/internal/extensions/patches/variables as those structs become available across the code base i.e. public or top-level internal.
+var builtinVariables = sets.NewString(
+	"builtin",
+
+	// Cluster builtins.
+	"builtin.cluster",
+	"builtin.cluster.name",
+	"builtin.cluster.namespace",
+
+	// ClusterTopology builtins.
+	"builtin.cluster.topology",
+	"builtin.cluster.topology.version",
+	"builtin.cluster.topology.class",
+
+	// ControlPlane builtins.
+	"builtin.controlPlane",
+	"builtin.controlPlane.version",
+	"builtin.controlPlane.replicas",
+
+	// MachineDeployment builtins.
+	"builtin.machineDeployment",
+	"builtin.machineDeployment.version",
+	"builtin.machineDeployment.class",
+	"builtin.machineDeployment.name",
+	"builtin.machineDeployment.topologyName",
+	"builtin.machineDeployment.replicas",
+)
+
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
+	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
+	return variablesMap
+}
+
+// validateIndexAccess checks to see if the jsonPath is attempting to add an element in the array i.e. access by number
+// If the operation is add an error is thrown if a number greater than 0 is used as an index.
+// If the operation is replace an error is thrown if an index is used.
+func validateIndexAccess(jsonPatch clusterv1.JSONPatch, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	pathParts := strings.Split(jsonPatch.Path, "/")
+	for _, part := range pathParts {
+		// If the jsonPatch operation is add disallow any number greater than 0. 0 here is a prepend operation on the array.
+		index, err := strconv.Atoi(part)
+		if err != nil {
+			continue
+		}
+
+		if jsonPatch.Op == "add" && index != 0 {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					"arrays can only be accessed using [0] (prepend) or [-] (append)",
+				))
+		}
+
+		// If the jsonPatch operation is replace or remove disallow any number as an element in the path.
+		if jsonPatch.Op == "replace" || jsonPatch.Op == "remove" {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					fmt.Sprintf("elements in arrays can not be accessed in a %s operation", jsonPatch.Op),
+				))
+		}
+	}
+	return allErrs
+}
+
+func prettyPrint(v interface{}) string {
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err != nil {
+		return err.Error()
```

Comment:
Nice way to avoid returning an error here, I like it!

Should we maybe wrap the error, because otherwise it can be quite obscure

```suggestion
		return errors.Wrapf(err, "failed to marshal value").Error()
```

Reply:
I'm not taking credit for this one - this code is taken directly from the testing framework (didn't want to import the testing package here) and I've seen similar in a lot of places 😄